### ISSUE TITLE
reading on uninitialized buffer can cause UB (`impl<R> BufRead for GreedyAccessReader<R>`)

### ISSUE THREAD

**Issue by JOE1994** [Enet4/bra-rs#1]
Hello :crab: ,
we (Rust group @sslab-gatech) found a memory-safety/soundness issue in this crate while scanning Rust code on crates.io for potential vulnerabilities.

## Issue Description
https://github.com/Enet4/bra-rs/blob/2b3c455fb3563e8259458275b4f6fd245dd69c47/src/greedy.rs#L190-L225

`GreedyAccessReader::fill_buf` method creates an uninitialized buffer and passes it to user-provided `Read` implementation (`self.inner.read(buf)`). This is unsound, because it allows safe Rust code to exhibit an **undefined behavior** (read from uninitialized memory).

[This part](https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read) from the `Read` trait documentation explains the issue:

> It is your responsibility to make sure that `buf` is initialized before calling `read`. Calling read with an uninitialized `buf` (of the kind one obtains via `MaybeUninit<T>`) is not safe, and can lead to undefined behavior.

## Suggested Fix
It is safe to zero-initialize the newly allocated `u8` buffer before `read()`, in order to prevent user-provided `Read` from reading old contents of the newly allocated heap memory.

The version available on `Crates.io` seems to be different from the latest master branch of this repo,
but the same issue exists in `GreedyBufReader::fill_buf()` (bra = "0.1.0").

Thank you for checking out this issue :+1:

**Comment by Enet4** [comment]
Hello, and thank you for the detailed report!

It is a bit surprising to find out that the documentation for `read` was expanded to explicitly require the slice to be initialized. The paragraph quoted from the documentation did not even exist back when I implemented this (the associated [commit](https://github.com/rust-lang/rust/commit/1c12b1be330dd9c3de0b4fe599686d7c0c45e720) was released with Rust 1.37.0). Still, it is understandable why this has to be done, considering that there is no `unsafe` construct barring someone from making a poorly behaved impl of `Read`.

Although this is a very low-traffic crate, I will take the necessary measures to fix this and release a patched version. :+1:
